tests/device_selector/device_selector_aspect.cpp

```diff
@@ -395,6 +396,49 @@ constexpr void check_for_random_set(const named_type_pack<AspectsT...>&) {
   create_random_sets(aspect_tuple, set_sizes);
 }
 
+template <typename AspectT>
+class kernel_aspect;
+#if !SYCL_CTS_COMPILING_WITH_COMPUTECPP
+/**
+ Functor that checks any_device_has and all_devices_have functionality. */
+template <typename AspectT>
+class check_any_device_has_all_devices_have {
+  static constexpr sycl::aspect aspect = AspectT::value;
+
+ public:
+  void operator()(const std::string& aspect_name) {
+    using k_name = kernel_aspect<AspectT>;
+
+    sycl::queue queue = util::get_cts_object::queue();
+    queue.submit(
+        [&](sycl::handler& cgh) { cgh.single_task<k_name>([=]() {}); });
+
+    auto exec_device_has_aspect = [=](auto d) {
+      return d.has(aspect) && sycl::is_compatible<k_name>(d);
+    };
+    auto devices = sycl::device::get_devices();
+    bool any_device =
+        any_of(devices.begin(), devices.end(), exec_device_has_aspect);
+    bool all_devices =
+        all_of(devices.begin(), devices.end(), exec_device_has_aspect);
+    if (any_device) {
+      INFO(
+          "Check that if some device has aspect A, "
+          "any_device_has_v<A> is true.");
+      CHECK(std::is_base_of_v<std::true_type, sycl::any_device_has<aspect>>);
+      CHECK(sycl::any_device_has_v<aspect>);
+    }
```

Comment:
`else` branch is missing, but I think we can simplify the code by checking that `sycl::any_device_has_v<aspect>` value matches `any_device`.

```c++
CHECK(sycl::any_device_has_v<aspect> == any_device);
```

The same applies to `all_devices`.

Reply:
The `else` case is not there because according to Spec "The trait any_device_has<aspect> inherits from std::true_type if the compilation environment supports any device which has the specified aspect" but it is possible that the compilation environment could support some devices that are not present on the system where test is run. So even when any_device is `false` we can't be sure that there are no such devices supported by the compilation environment and that any_device_has_v is supposed to be `false`. As discussed here https://github.com/KhronosGroup/SYCL-CTS/pull/539#discussion_r1114877847.
Same with `all_devices`.

Added comments in test to clarify that.

Reply:
I need to confess; I hadn't read the spec before I reviewed this patch the first time. Now I think this test doesn't work as intended.
1. **Coverage**. The spec says "The trait any_device_has<aspect> inherits from std::true_type if the compilation environment supports any device which has the specified aspect, and it inherits from std::false_type if no device has the aspect.". The test attempts to check the first part of this statement, but not the second one and I still believe that we can check the second part too.
2. **Compile-time traits vs runtime**. Line 419 returns **_all_** devices the test finds in the system at runtime, but it can be a superset of device we compiled the test for. Let's say we compiled for a x86 CPU device but run the test on a system with x86 CPU and NVIDIA GPU device. Even if we run the test on CPU device, the runtime query `any_device` will be `true` for the aspects supported by GPU device, but not supported by CPU device. This is a bug in the test. I think we can fix it by quiring the devices associated with `kernel_bundle` rather than all available devices.
3. **JIT compilation**. How is this supposed to work for "runtime" compilation environments? I think by "compilation environment" specification means compile-time when we build the test and at the same time specification allows implementations to use "online compilation". When implementation supports online compilation, the device might be unknown in "compilation environment". For example, DPC++ uses SPIR-V (i.e. intermediate language) target with all possible aspects enabled by default, so I would expect all compile-time traits will return `true` (tagging @AlexeySachkov and @steffenlarsen to confirm). I see no way for runtime query to return a "virtual" device supporting all aspects.

Reply:
1. It may be possible to check the second part, but it would be the other way around from the current tests. I imagine something like "if `any_device_has<aspect>` inherits from `std::false_type`, check that all devices return `false` for `has(aspect)`". Is that what you had in mind?
2. I believe that `sycl::is_compatible` was intended to address this, but reading the specification it does seem like "compatibility" only considers the aspects and `device_has` attributes (as well as builtin kernel ownership). Should this also consider the compilation environment (tag @gmlueck)? Otherwise, I agree we would need to use `kernel_bundle`, but I am not even convinced it necessarily makes any stronger guarantees than `is_compatible`.
3. These traits are arguably not very useful for more open compilation environments, such as the SPIR-V JIT case. Depending on the environment, you may be able to make some guarantees, but in the most general case you would have `any_device_has` being trivially true and `all_devices_have` being trivially false.

Reply:
> I agree that we should make a spec clarification saying that `is_compatible` returns `false` when the translation unit containing the kernel does not support the given device. I can make this clarification.

I created this PR: KhronosGroup/SYCL-Docs#381